File: classifier/learning.py
"""
Preference learning via Ridge regression with interaction features.

User rates historical matches 1–10.  We fit a Ridge regression on raw scorer
values plus pairwise interaction terms (raw_i × raw_j).  Positive coefficients
→ higher weight.  Blend data-driven weights with DEFAULT_WEIGHTS using a
confidence ramp: at few ratings the prior dominates, at ~20+ the data takes over.

Active learning: when selecting calibration matches, pick those where the
current model is most uncertain (highest prediction variance via bootstrap).
"""
from __future__ import annotations

from itertools import combinations

import numpy as np

from .models import Match, UserProfile

# Must match the scorers in build_default_engine() exactly.
SCORER_NAMES: list[str] = [
    "Favorite Team",
    "Competitive Tension",
    "Match Stage",
    "Star Power",
    "Chaos Potential",
    "Form",
    "Narrative",
    "Same Group",
]

# Key interaction pairs — discovered combos that matter
INTERACTION_PAIRS: list[tuple[str, str]] = [
    ("Favorite Team", "Match Stage"),       # your team in a big match
    ("Competitive Tension", "Star Power"),   # close game between stars
    ("Chaos Potential", "Match Stage"),      # chaos in knockout = drama
    ("Competitive Tension", "Narrative"),    # rivalry + close = electric
    ("Favorite Team", "Star Power"),         # your team has stars
    ("Chaos Potential", "Narrative"),        # chaotic rivalry
    ("Competitive Tension", "Chaos Potential"),  # Vecer (2007): close + high-scoring = excitement
]

INTERACTION_NAMES: list[str] = [f"{a} × {b}" for a, b in INTERACTION_PAIRS]

DEFAULT_WEIGHTS: dict[str, float] = {
    "Favorite Team":        0.19,
    "Competitive Tension":  0.18,
    "Match Stage":          0.17,
    "Star Power":           0.17,
    "Chaos Potential":      0.12,
    "Form":                 0.08,
    "Narrative":            0.06,
    "Same Group":           0.03,
}

_RAMP_N = 20  # ratings needed for full data confidence (down from 30 — Ridge needs less)


def rating_to_label(r: float) -> str:
    """Map a 1–10 mean rating to a category label."""
    if r >= 7.5:
        return "🔥 Imperdible"
    if r >= 4.5:
        return "👀 Tal Vez"
    return "📺 Resumen"


# ── Feature matrix construction ──────────────────────────────────────────────

def _build_feature_matrix(
    raw_examples: list[dict],
) -> tuple[np.ndarray, list[str]]:
    """
    Build feature matrix with base scorers + interaction terms.
    Returns (X, feature_names).
    """
    n = len(raw_examples)
    n_base = len(SCORER_NAMES)
    n_interact = len(INTERACTION_PAIRS)
    feature_names = list(SCORER_NAMES) + list(INTERACTION_NAMES)

    X = np.zeros((n, n_base + n_interact), dtype=float)

    for i, ex in enumerate(raw_examples):
        raw = ex["raw"]
        # Base features
        for j, s in enumerate(SCORER_NAMES):
            X[i, j] = raw.get(s, 0.5)
        # Interaction features
        for k, (a, b) in enumerate(INTERACTION_PAIRS):
            X[i, n_base + k] = raw.get(a, 0.5) * raw.get(b, 0.5)

    return X, feature_names


def _build_single_vector(raw: dict[str, float]) -> np.ndarray:
    """Build feature vector for a single match (for prediction/uncertainty)."""
    n_base = len(SCORER_NAMES)
    n_interact = len(INTERACTION_PAIRS)
    vec = np.zeros(n_base + n_interact, dtype=float)

    for j, s in enumerate(SCORER_NAMES):
        vec[j] = raw.get(s, 0.5)
    for k, (a, b) in enumerate(INTERACTION_PAIRS):
        vec[n_base + k] = raw.get(a, 0.5) * raw.get(b, 0.5)

    return vec


# ── Ridge regression weights ────────────────────────────────────────────────

def _ridge_weights(X: np.ndarray, y: np.ndarray) -> dict[str, float]:
    """
    Fit Ridge regression, extract base-scorer weights from coefficients.
    Interaction coefficients are folded into the model but reported separately.
    """
    from sklearn.linear_model import Ridge

    model = Ridge(alpha=1.0, fit_intercept=True)
    model.fit(X, y)

    coefs = model.coef_
    n_base = len(SCORER_NAMES)

    # Base scorer weights: use abs(coef) as importance, then normalise
    base_coefs = coefs[:n_base]
    importance = np.array([max(0.0, c) for c in base_coefs])

    if importance.sum() < 1e-9:
        return {name: 1.0 / n_base for name in SCORER_NAMES}

    normed = importance / importance.sum()
    return {name: round(float(v), 4) for name, v in zip(SCORER_NAMES, normed)}


def _ridge_full(X: np.ndarray, y: np.ndarray) -> tuple[dict[str, float], dict[str, float], "Ridge"]:
    """
    Full Ridge fit returning base weights, interaction strengths, and the model.
    """
    from sklearn.linear_model import Ridge

    model = Ridge(alpha=1.0, fit_intercept=True)
    model.fit(X, y)

    coefs = model.coef_
    n_base = len(SCORER_NAMES)

    # Base weights
    base_coefs = coefs[:n_base]
    importance = np.array([max(0.0, c) for c in base_coefs])
    if importance.sum() < 1e-9:
        base_weights = {name: 1.0 / n_base for name in SCORER_NAMES}
    else:
        normed = importance / importance.sum()
        base_weights = {name: round(float(v), 4) for name, v in zip(SCORER_NAMES, normed)}

    # Interaction strengths (raw coefficients, not normalised)
    interact_coefs = coefs[n_base:]
    interactions = {
        name: round(float(c), 4)
        for name, c in zip(INTERACTION_NAMES, interact_coefs)
    }

    return base_weights, interactions, model


# ── Bayesian blend ──────────────────────────────────────────────────────────

def _blend(
    prior: dict[str, float],
    data: dict[str, float],
    n_ratings: int,
) -> dict[str, float]:
    """
    Bayesian blend: (1 - α) * prior + α * data.
    α ramps linearly from 0 to 1 over _RAMP_N ratings.
    """
    alpha = min(1.0, n_ratings / _RAMP_N)
    blended = {}
    for name in SCORER_NAMES:
        p = prior.get(name, 0.0)
        d = data.get(name, 0.0)
        blended[name] = (1 - alpha) * p + alpha * d

    total = sum(blended.values())
    if total > 0:
        blended = {k: round(v / total, 4) for k, v in blended.items()}
    return blended


# ── Active learning: uncertainty estimation ─────────────────────────────────

def predict_uncertainty(
    rated_examples: list[dict],
    candidate_raws: list[dict[str, float]],
    n_bootstrap: int = 50,
) -> list[float]:
    """
    Estimate prediction uncertainty for candidate matches using bootstrap.

    Fits n_bootstrap Ridge models on resampled training data, predicts each
    candidate, returns std of predictions as uncertainty measure.

    Higher uncertainty = model disagrees across bootstraps = more informative
    to rate this match next.
    """
    if len(rated_examples) < 5:
        # Not enough data — return uniform uncertainty (feature spread as proxy)
        uncertainties = []
        for raw in candidate_raws:
            vec = _build_single_vector(raw)
            spread = float(np.std(vec))
            uncertainties.append(spread)
        return uncertainties

    X_train, _ = _build_feature_matrix(rated_examples)
    y_train = np.array([float(ex["rating"]) for ex in rated_examples])

    X_cand = np.array([_build_single_vector(raw) for raw in candidate_raws])

    from sklearn.linear_model import Ridge

    rng = np.random.RandomState(42)
    predictions = np.zeros((n_bootstrap, len(candidate_raws)))

    for b in range(n_bootstrap):
        idx = rng.choice(len(y_train), size=len(y_train), replace=True)
        model = Ridge(alpha=1.0, fit_intercept=True)
        model.fit(X_train[idx], y_train[idx])
        predictions[b] = model.predict(X_cand)

    return [float(v) for v in np.std(predictions, axis=0)]


# ── Public API ────────────────────────────────────────────────────────────────

def fit_from_ratings(rated_examples: list[dict]) -> dict:
    """
    Learn weights from (raw_score_vector, rating 1–10) examples.

    Uses Ridge regression with interaction features when enough data exists,
    falls back to Pearson correlation for < 5 examples.

    Returns:
      {weights, weight_delta, top_features, interactions, rating_stats, confidence, method}
    """
    if not rated_examples:
        return {
            "weights":      dict(DEFAULT_WEIGHTS),
            "weight_delta": {n: 0.0 for n in SCORER_NAMES},
            "top_features": [],
            "interactions": {},
            "rating_stats": {},
            "confidence":   0.0,
            "method":       "prior",
        }

    X, feature_names = _build_feature_matrix(rated_examples)
    y = np.array([float(ex["rating"]) for ex in rated_examples], dtype=float)

    n_samples = len(y)

    if n_samples < 5:
        # Too few for Ridge — use simple Pearson on base features only
        data_weights = _pearson_weights(X[:, :len(SCORER_NAMES)], y)
        weights = _blend(DEFAULT_WEIGHTS, data_weights, n_samples)
        interactions = {}
        method = "pearson"
    else:
        data_weights, interactions, model = _ridge_full(X, y)
        weights = _blend(DEFAULT_WEIGHTS, data_weights, n_samples)
        method = "ridge"

    top_features = sorted(
        [{"scorer": name, "importance": round(float(v), 4)}
         for name, v in weights.items()],
        key=lambda x: x["importance"],
        reverse=True,
    )

    # Top interactions (only significant ones)
    top_interactions = {
        k: v for k, v in sorted(interactions.items(), key=lambda x: abs(x[1]), reverse=True)
        if abs(v) > 0.1
    }

    rating_stats = {
        "mean": round(float(np.mean(y)), 1),
        "min":  int(np.min(y)),
        "max":  int(np.max(y)),
        "n":    len(y),
        "dist": {str(i): int((y == i).sum()) for i in range(1, 11) if (y == i).sum() > 0},
    }

    weight_delta = {
        name: round(float(weights[name]) - DEFAULT_WEIGHTS.get(name, 0.0), 4)
        for name in SCORER_NAMES
    }

    confidence = round(min(1.0, len(y) / _RAMP_N), 2)

    return {
        "weights":      weights,
        "weight_delta": weight_delta,
        "top_features": top_features[:6],
        "interactions": top_interactions,
        "rating_stats": rating_stats,
        "confidence":   confidence,
        "method":       method,
    }


# ── Fallback: Pearson for < 5 samples ───────────────────────────────────────

def _pearson(x: np.ndarray, y: np.ndarray) -> float:
    if len(x) < 3:
        return 0.0
    sx, sy = float(np.std(x)), float(np.std(y))
    if sx < 1e-9 or sy < 1e-9:
        return 0.0
    return float(np.corrcoef(x, y)[0, 1])


def _pearson_weights(X: np.ndarray, y: np.ndarray) -> dict[str, float]:
    n_scorers = X.shape[1]
    raw = np.array([max(0.0, _pearson(X[:, i], y)) for i in range(n_scorers)])
    if raw.sum() < 1e-9:
        return {name: 1.0 / n_scorers for name in SCORER_NAMES}
    normed = raw / raw.sum()
    return {name: round(float(v), 4) for name, v in zip(SCORER_NAMES, normed)}

File: frontend/src/scoring/learning.ts
/**
 * Preference learning — TypeScript port of classifier/learning.py.
 *
 * Ridge regression with interaction features, Pearson fallback for < 5 samples,
 * Bayesian blend with default weights using confidence ramp.
 *
 * No external dependencies — matrix math is inlined.
 */

// ── Constants ────────────────────────────────────────────────────────────────

export const SCORER_NAMES = [
  "Favorite Team",
  "Competitive Tension",
  "Match Stage",
  "Star Power",
  "Chaos Potential",
  "Form",
  "Narrative",
  "Same Group",
] as const;

const INTERACTION_PAIRS: [string, string][] = [
  ["Favorite Team", "Match Stage"],
  ["Competitive Tension", "Star Power"],
  ["Chaos Potential", "Match Stage"],
  ["Competitive Tension", "Narrative"],
  ["Favorite Team", "Star Power"],
  ["Chaos Potential", "Narrative"],
  ["Competitive Tension", "Chaos Potential"],
];

const INTERACTION_NAMES = INTERACTION_PAIRS.map(([a, b]) => `${a} × ${b}`);

export const DEFAULT_WEIGHTS: Record<string, number> = {
  "Favorite Team": 0.19,
  "Competitive Tension": 0.18,
  "Match Stage": 0.17,
  "Star Power": 0.17,
  "Chaos Potential": 0.12,
  "Form": 0.08,
  "Narrative": 0.06,
  "Same Group": 0.03,
};

const RAMP_N = 20;
const N_BASE = SCORER_NAMES.length;
const N_INTERACT = INTERACTION_PAIRS.length;
// const N_FEATURES = N_BASE + N_INTERACT;

// ── Types ────────────────────────────────────────────────────────────────────

export interface RatedExample {
  raw: Record<string, number>;
  rating: number;
}

export interface TopFeature {
  scorer: string;
  importance: number;
}

export interface RatingStats {
  mean: number;
  min: number;
  max: number;
  n: number;
  dist: Record<string, number>;
}

export interface FitResult {
  weights: Record<string, number>;
  weight_delta: Record<string, number>;
  top_features: TopFeature[];
  interactions: Record<string, number>;
  rating_stats: RatingStats;
  confidence: number;
  method: "prior" | "pearson" | "ridge";
  scorer_labels?: Record<string, string>;
  total_examples?: number;
}

// ── Feature matrix ───────────────────────────────────────────────────────────

function buildFeatureRow(raw: Record<string, number>): number[] {
  const row: number[] = [];
  for (const s of SCORER_NAMES) row.push(raw[s] ?? 0.5);
  for (const [a, b] of INTERACTION_PAIRS) row.push((raw[a] ?? 0.5) * (raw[b] ?? 0.5));
  return row;
}

function buildFeatureMatrix(examples: RatedExample[]): { X: number[][]; y: number[] } {
  const X = examples.map((ex) => buildFeatureRow(ex.raw));
  const y = examples.map((ex) => ex.rating);
  return { X, y };
}

// ── Matrix math (small matrices only) ────────────────────────────────────────

function transpose(A: number[][]): number[][] {
  const rows = A.length, cols = A[0].length;
  const T: number[][] = Array.from({ length: cols }, () => new Array(rows));
  for (let i = 0; i < rows; i++)
    for (let j = 0; j < cols; j++)
      T[j][i] = A[i][j];
  return T;
}

function matMul(A: number[][], B: number[][]): number[][] {
  const m = A.length, n = B[0].length, p = B.length;
  const C: number[][] = Array.from({ length: m }, () => new Array(n).fill(0));
  for (let i = 0; i < m; i++)
    for (let j = 0; j < n; j++)
      for (let k = 0; k < p; k++)
        C[i][j] += A[i][k] * B[k][j];
  return C;
}

function matVecMul(A: number[][], v: number[]): number[] {
  return A.map((row) => row.reduce((s, a, i) => s + a * v[i], 0));
}

/** Gauss-Jordan inversion for small square matrices. */
function invert(M: number[][]): number[][] {
  const n = M.length;
  // Augment [M | I]
  const A = M.map((row, i) => {
    const aug = new Array(2 * n).fill(0);
    for (let j = 0; j < n; j++) aug[j] = row[j];
    aug[n + i] = 1;
    return aug;
  });

  for (let col = 0; col < n; col++) {
    // Partial pivot
    let maxRow = col;
    for (let r = col + 1; r < n; r++)
      if (Math.abs(A[r][col]) > Math.abs(A[maxRow][col])) maxRow = r;
    [A[col], A[maxRow]] = [A[maxRow], A[col]];

    const pivot = A[col][col];
    if (Math.abs(pivot) < 1e-12) {
      // Singular — add regularization and retry shouldn't happen with Ridge
      A[col][col] += 1e-6;
    }
    const inv = 1 / A[col][col];
    for (let j = 0; j < 2 * n; j++) A[col][j] *= inv;

    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = A[r][col];
      for (let j = 0; j < 2 * n; j++) A[r][j] -= factor * A[col][j];
    }
  }

  return A.map((row) => row.slice(n));
}

// ── Ridge regression ────────────────────────────────────────────────────────

function ridgeFit(
  X: number[][],
  y: number[],
  alpha = 1.0,
): { coefs: number[]; intercept: number } {
  const n = X.length;
  const p = X[0].length;

  // Center y
  const yMean = y.reduce((s, v) => s + v, 0) / n;
  const yc = y.map((v) => v - yMean);

  // Center X columns
  const colMeans = new Array(p).fill(0);
  for (let j = 0; j < p; j++) {
    for (let i = 0; i < n; i++) colMeans[j] += X[i][j];
    colMeans[j] /= n;
  }
  const Xc = X.map((row) => row.map((v, j) => v - colMeans[j]));

  // XtX + αI
  const Xt = transpose(Xc);
  const XtX = matMul(Xt, Xc);
  for (let i = 0; i < p; i++) XtX[i][i] += alpha;

  // Xty
  const Xty = matVecMul(Xt, yc);

  // Solve: coefs = (XtX)^-1 Xty
  const XtXinv = invert(XtX);
  const coefs = matVecMul(XtXinv, Xty);

  // Intercept
  const intercept = yMean - colMeans.reduce((s, m, j) => s + m * coefs[j], 0);

  return { coefs, intercept };
}

function ridgePredict(
  X: number[][],
  coefs: number[],
  intercept: number,
): number[] {
  return X.map((row) => row.reduce((s, v, j) => s + v * coefs[j], 0) + intercept);
}

// ── Extract weights from Ridge ──────────────────────────────────────────────

function ridgeToWeights(coefs: number[]): Record<string, number> {
  const baseCoefs = coefs.slice(0, N_BASE);
  const importance = baseCoefs.map((c) => Math.max(0, c));
  const sum = importance.reduce((s, v) => s + v, 0);

  if (sum < 1e-9) {
    return Object.fromEntries(SCORER_NAMES.map((n) => [n, 1 / N_BASE]));
  }

  return Object.fromEntries(
    SCORER_NAMES.map((n, i) => [n, Math.round((importance[i] / sum) * 10000) / 10000]),
  );
}

function ridgeToInteractions(coefs: number[]): Record<string, number> {
  const interCoefs = coefs.slice(N_BASE);
  const result: Record<string, number> = {};
  for (let i = 0; i < N_INTERACT; i++) {
    const v = Math.round(interCoefs[i] * 10000) / 10000;
    if (Math.abs(v) > 0.1) result[INTERACTION_NAMES[i]] = v;
  }
  return result;
}

// ── Pearson correlation ─────────────────────────────────────────────────────

function pearson(x: number[], y: number[]): number {
  const n = x.length;
  if (n < 3) return 0;
  const mx = x.reduce((s, v) => s + v, 0) / n;
  const my = y.reduce((s, v) => s + v, 0) / n;
  let num = 0, dx2 = 0, dy2 = 0;
  for (let i = 0; i < n; i++) {
    const dx = x[i] - mx, dy = y[i] - my;
    num += dx * dy;
    dx2 += dx * dx;
    dy2 += dy * dy;
  }
  const denom = Math.sqrt(dx2 * dy2);
  return denom < 1e-9 ? 0 : num / denom;
}

function pearsonWeights(X: number[][], y: number[]): Record<string, number> {
  const raw = SCORER_NAMES.map((_, i) => Math.max(0, pearson(X.map((r) => r[i]), y)));
  const sum = raw.reduce((s, v) => s + v, 0);
  if (sum < 1e-9) return Object.fromEntries(SCORER_NAMES.map((n) => [n, 1 / N_BASE]));
  return Object.fromEntries(SCORER_NAMES.map((n, i) => [n, Math.round((raw[i] / sum) * 10000) / 10000]));
}

// ── Bayesian blend ──────────────────────────────────────────────────────────

function blend(
  prior: Record<string, number>,
  data: Record<string, number>,
  nRatings: number,
): Record<string, number> {
  const alpha = Math.min(1.0, nRatings / RAMP_N);
  const blended: Record<string, number> = {};
  for (const name of SCORER_NAMES) {
    blended[name] = (1 - alpha) * (prior[name] ?? 0) + alpha * (data[name] ?? 0);
  }
  const total = Object.values(blended).reduce((s, v) => s + v, 0);
  if (total > 0) {
    for (const name of SCORER_NAMES) blended[name] = Math.round((blended[name] / total) * 10000) / 10000;
  }
  return blended;
}

// ── Active learning: uncertainty estimation ─────────────────────────────────

export function predictUncertainty(
  ratedExamples: RatedExample[],
  candidateRaws: Record<string, number>[],
  nBootstrap = 50,
): number[] {
  if (ratedExamples.length < 5) {
    return candidateRaws.map((raw) => {
      const vec = buildFeatureRow(raw);
      const mean = vec.reduce((s, v) => s + v, 0) / vec.length;
      return Math.sqrt(vec.reduce((s, v) => s + (v - mean) ** 2, 0) / vec.length);
    });
  }

  const { X: Xtrain, y: ytrain } = buildFeatureMatrix(ratedExamples);
  const Xcand = candidateRaws.map(buildFeatureRow);
  const n = ytrain.length;

  // Simple seeded random for reproducibility
  let seed = 42;
  const rand = () => { seed = (seed * 1664525 + 1013904223) & 0x7fffffff; return seed / 0x7fffffff; };

  const predictions: number[][] = Array.from({ length: Xcand.length }, () => []);

  for (let b = 0; b < nBootstrap; b++) {
    const idx = Array.from({ length: n }, () => Math.floor(rand() * n));
    const Xb = idx.map((i) => Xtrain[i]);
    const yb = idx.map((i) => ytrain[i]);
    const { coefs, intercept } = ridgeFit(Xb, yb);
    const preds = ridgePredict(Xcand, coefs, intercept);
    for (let j = 0; j < Xcand.length; j++) predictions[j].push(preds[j]);
  }

  return predictions.map((preds) => {
    const mean = preds.reduce((s, v) => s + v, 0) / preds.length;
    return Math.sqrt(preds.reduce((s, v) => s + (v - mean) ** 2, 0) / preds.length);
  });
}

// ── Public API ──────────────────────────────────────────────────────────────

export function fitFromRatings(ratedExamples: RatedExample[]): FitResult {
  if (!ratedExamples.length) {
    return {
      weights: { ...DEFAULT_WEIGHTS },
      weight_delta: Object.fromEntries(SCORER_NAMES.map((n) => [n, 0])),
      top_features: [],
      interactions: {},
      rating_stats: { mean: 0, min: 0, max: 0, n: 0, dist: {} },
      confidence: 0,
      method: "prior",
    };
  }

  const { X, y } = buildFeatureMatrix(ratedExamples);
  const nSamples = y.length;

  let dataWeights: Record<string, number>;
  let interactions: Record<string, number> = {};
  let method: "pearson" | "ridge";

  if (nSamples < 5) {
    dataWeights = pearsonWeights(X, y);
    method = "pearson";
  } else {
    const { coefs } = ridgeFit(X, y);
    dataWeights = ridgeToWeights(coefs);
    interactions = ridgeToInteractions(coefs);
    method = "ridge";
  }

  const weights = blend(DEFAULT_WEIGHTS, dataWeights, nSamples);

  const topFeatures = Object.entries(weights)
    .map(([scorer, importance]) => ({ scorer, importance: Math.round(importance * 10000) / 10000 }))
    .sort((a, b) => b.importance - a.importance)
    .slice(0, 6);

  const topInteractions = Object.fromEntries(
    Object.entries(interactions)
      .filter(([, v]) => Math.abs(v) > 0.1)
      .sort(([, a], [, b]) => Math.abs(b) - Math.abs(a)),
  );

  const mean = y.reduce((s, v) => s + v, 0) / y.length;
  const dist: Record<string, number> = {};
  for (const r of y) {
    const k = String(Math.round(r));
    dist[k] = (dist[k] ?? 0) + 1;
  }

  return {
    weights,
    weight_delta: Object.fromEntries(
      SCORER_NAMES.map((n) => [n, Math.round((weights[n] - (DEFAULT_WEIGHTS[n] ?? 0)) * 10000) / 10000]),
    ),
    top_features: topFeatures,
    interactions: topInteractions,
    rating_stats: {
      mean: Math.round(mean * 10) / 10,
      min: Math.min(...y),
      max: Math.max(...y),
      n: y.length,
      dist,
    },
    confidence: Math.round(Math.min(1.0, y.length / RAMP_N) * 100) / 100,
    method,
  };
}
